### PR TITLE
docs: add a proposal for distributed DDL reorg

### PR DESCRIPTION
**Pull request by zimulala**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/37129

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [x] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: docs/design/2022-09-19-distributed-ddl-reorg.md [hunk 1]
```diff
@@ -0,0 +1,310 @@
+# Proposal: Distributed DDL Reorg
+
+- Author(s): [zimulala](https://github.com/zimulala), [Defined2014](https://github.com/Defined2014)
+- Tracking Issue: https://github.com/pingcap/tidb/issues/41208
+
+# <a id="_37pie9l9cba0"></a>__Abstract__
```
Comment (Defined2014) [inline docs/design/2022-09-19-distributed-ddl-reorg.md:6]: Why add id here?

File: docs/design/2022-09-19-distributed-ddl-reorg.md [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+
+![Figure 2: dist reorg flow chart](./imgs/dist-reorg-flow-chart.png)
+
+## <a id="_3sqnjcjlr92v"></a>__Detailed Design__
+
+### <a id="_9bup4r9f5b8d"></a>__Meta Info Definition__
+
+The contents of the existing table structure may be lacking, and a new Metadata needs to be added or defined.
+
+Add fields to the DDLReorgMeta structure of the mysql.TiDB_ddl_job table as follows:
```
Comment (Defined2014) [inline docs/design/2022-09-19-distributed-ddl-reorg.md:59]:
```suggestion
Add fields to the DDLReorgMeta structure of the `mysql.tidb_ddl_job` table as follows:
```

File: docs/design/2022-09-19-distributed-ddl-reorg.md [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+
+```go
+type DDLReorgMeta struct {
+    ...                    // Some of the original fields
+    
+    IsDistReorg   bool     // Determine whether do dist-reorg
+}
+```
+
+Consider that if all subtask information is added to the TiDB_ddl_reorg.reorg field, there may be a lock  problem. It is added to the mysql. TiDB_background_subtask table, the specific structure is as follows:
```
Comment (Defined2014) [inline docs/design/2022-09-19-distributed-ddl-reorg.md:69]:
```suggestion
Consider that if all subtask information is added to the TiDB_ddl_reorg.reorg field, there may be a lock problem. It is added to the mysql.tidb_background_subtask table, the specific structure is as follows:
```

Do we have `tidb_background_subtask` table in v6.6.0?

File: docs/design/2022-09-19-distributed-ddl-reorg.md [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+### <a id="_p1kj0bi482xl"></a>__Communication Mode__
+
+In the current scheme, the backfill worker obtains subtasks and the reorg worker checks whether the subtask is completed through regular inspection and processing. Here, we consider combining PD watches for communication.
+
+### <a id="_q2j6xv14l488"></a>__Breakpoints Resume__
+
+When the network partition or abnormal exit occurs in the TiDB where the current backfill worker is located, the corresponding subtask may not be handled by the worker. In the current scheme, it is tentatively planned to mark whether the executor owner of the current subtask is valid by lease. There are more suitable schemes that can be discussed later. The specific operation of this scheme:
+
+1. When the backfill worker handles a subtask, it will record the current DDL_ID (may need worker_type_worker_id suffix) in the TiDB_background_subtask table as the exec_id, and regularly update the exec_expired value and curr_key.
+2. Non- DDL owner TiDB encountered this problem:
```
Comment (Defined2014) [inline docs/design/2022-09-19-distributed-ddl-reorg.md:261]:
```suggestion
2. Non-DDL owner TiDB encountered this problem:
```

File: docs/design/2022-09-19-distributed-ddl-reorg.md [hunk 5]
```diff
@@ -0,0 +1,310 @@
+# Proposal: Distributed DDL Reorg
+
+- Author(s): [zimulala](https://github.com/zimulala), [Defined2014](https://github.com/Defined2014)
+- Tracking Issue: https://github.com/pingcap/tidb/issues/41208
+
+# <a id="_37pie9l9cba0"></a>__Abstract__
+
+This is distributed processing of design in the DDL reorg phase.
```
Comment (Benjamin2037) [inline docs/design/2022-09-19-distributed-ddl-reorg.md:8]: I do not think here need use two paragraphs
“The document proposes a design for the distributed processing of DDL reorg phase in TiDB. The current design is based on the main logic that only the DDL owner can handle DDL jobs. However, this proposal suggests that all TiDBs should be able to claim subtasks in the reorg phase based on resource usage.”

**Comment by zimulala** [comment]
PTLA @Defined2014 @Benjamin2037

File: docs/design/2022-09-19-distributed-ddl-reorg.md [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+
+```go
+type DDLReorgMeta struct {
+    ...                    // Some of the original fields
+    
+    IsDistReorg   bool     // Determine whether do dist-reorg
+}
+```
+
+Consider that if all subtask information is added to the TiDB_ddl_reorg.reorg field, there may be a lock problem. It is added to the mysql.tidb_background_subtask table, the specific structure is as follows:
```
Comment (Defined2014) [inline docs/design/2022-09-19-distributed-ddl-reorg.md:67]:
```suggestion
Consider that if all subtask information is added to the TiDB_ddl_reorg.reorg field, there may be a lock problem. It is added to the `mysql.tidb_background_subtask` table, the specific structure is as follows:
```

File: docs/design/2022-09-19-distributed-ddl-reorg.md [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+
+![Figure 2: dist reorg flow chart](./imgs/dist-reorg-flow-chart.png)
+
+##Detailed Design
+
+###Meta Info Definition
+
+The contents of the existing table structure may be lacking, and a new Metadata needs to be added or defined.
+
+Add fields to the DDLReorgMeta structure of the `mysql.TiDB_ddl_job` table as follows:
```
Comment (Defined2014) [inline docs/design/2022-09-19-distributed-ddl-reorg.md:57]: don't understand this line here. :(

File: docs/design/2022-09-19-distributed-ddl-reorg.md [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+    EndKey          kv.Key
+    EndInclude      bool    
+    ReorgTp         ReorgType
+    ...
+    
+    *JobMeta // parent job meta
+}
+```
+
+Add mysql.TiDB_background_subtask_history table to record completed (including failure status) subtasks. The table structure is the same as tidb_background_subtask . Considering the number of subtasks, some records of the history table are deleted regularly in the later stage.
```
Comment (Defined2014) [inline docs/design/2022-09-19-distributed-ddl-reorg.md:118]:
```suggestion
Add `mysql.tidb_background_subtask_history` table to record completed (including failure status) subtasks. The table structure is the same as tidb_background_subtask. Considering the number of subtasks, some records of the history table are deleted regularly in the later stage.
```

**Review by Defined2014 — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**